File: lib/widgets/root_scaffold.dart
import 'package:flutter/material.dart';

import '../components/icon.dart';
import '../theme/theme.dart';

enum RootTab {
  systems('Systems', 'hosts'),
  users('Users', 'identities'),
  sessions('Sessions', 'terminal'),
  settings('Settings', 'settings');

  const RootTab(this.label, this.icon);

  final String label;
  final String icon;

  String asset(bool selected) {
    const hasFilled = {'hosts', 'identities', 'settings'};
    if (selected && hasFilled.contains(icon)) {
      return 'assets/ic/nav/$icon-filled.svg';
    }
    return 'assets/ic/nav/$icon.svg';
  }
}

class RootScaffold extends StatelessWidget {
  const RootScaffold({
    super.key,
    required this.child,
    required this.currentTab,
    required this.onTabSelected,
    this.sessionCount = 0,
  });

  final Widget child;
  final RootTab currentTab;
  final ValueChanged<RootTab> onTabSelected;
  final int sessionCount;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    return Scaffold(
      backgroundColor: colors.background,
      body: child,
      bottomNavigationBar: Container(
        color: colors.card,
        child: SafeArea(
          top: false,
          child: Padding(
            padding: const EdgeInsets.only(top: 6, bottom: 8),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                for (final tab in RootTab.values)
                  _BottomTab(
                    tab: tab,
                    selected: currentTab == tab,
                    badge: tab == RootTab.sessions ? sessionCount : 0,
                    onTap: () => onTabSelected(tab),
                  ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class _BottomTab extends StatelessWidget {
  const _BottomTab({
    required this.tab,
    required this.selected,
    required this.badge,
    required this.onTap,
  });

  final RootTab tab;
  final bool selected;
  final int badge;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    final color = selected ? colors.accent : colors.textMuted;
    return Material(
      color: Colors.transparent,
      borderRadius: BorderRadius.circular(10),
      clipBehavior: Clip.antiAlias,
      child: InkWell(
        borderRadius: BorderRadius.circular(10),
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              SizedBox(
                width: 42,
                height: 24,
                child: Stack(
                  alignment: Alignment.center,
                  clipBehavior: Clip.none,
                  children: [
                    QIcon(asset: tab.asset(selected), color: color, size: 24),
                    if (badge > 0)
                      Positioned(
                        right: 2,
                        top: -2,
                        child: Container(
                          padding: const EdgeInsets.symmetric(
                            horizontal: 5,
                            vertical: 1,
                          ),
                          decoration: BoxDecoration(
                            color: colors.accent,
                            borderRadius: BorderRadius.circular(8),
                          ),
                          child: Text(
                            '$badge',
                            style: TextStyle(
                              fontSize: 9,
                              height: 1.2,
                              fontWeight: FontWeight.w700,
                              color: colors.onAccent,
                            ),
                          ),
                        ),
                      ),
                  ],
                ),
              ),
              const SizedBox(height: 4),
              Text(
                tab.label,
                style: TextStyle(
                  fontSize: 10,
                  fontWeight: FontWeight.w700,
                  color: color,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/widgets/RootScaffold.kt
package com.example.ohmyssh.widgets

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.consumeWindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBars
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Dns
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Terminal
import androidx.compose.material.icons.outlined.Dns
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material.icons.outlined.Terminal
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.ohmyssh.theme.appColors

enum class RootTab(
    val label: String,
    val icon: ImageVector,
    val selectedIcon: ImageVector,
) {
    SYSTEMS("Systems", Icons.Outlined.Dns, Icons.Filled.Dns),
    USERS("Users", Icons.Outlined.Person, Icons.Filled.Person),
    SESSIONS("Sessions", Icons.Outlined.Terminal, Icons.Filled.Terminal),
    SETTINGS("Settings", Icons.Outlined.Settings, Icons.Filled.Settings),
}

@Composable
fun RootScaffold(
    currentTab: RootTab,
    onTabSelected: (RootTab) -> Unit,
    sessionCount: Int = 0,
    content: @Composable () -> Unit,
) {
    val colors = appColors
    Column(Modifier.fillMaxSize().background(colors.background)) {
        // The tab bar already covers the navigation-bar inset; without
        // consuming it here every tab page would pad the bottom twice.
        Box(
            Modifier
                .weight(1f)
                .fillMaxWidth()
                .consumeWindowInsets(WindowInsets.navigationBars),
        ) { content() }
        Column(
            Modifier
                .fillMaxWidth()
                .background(colors.card)
                .windowInsetsPadding(WindowInsets.navigationBars),
        ) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(top = 6.dp, bottom = 8.dp),
                horizontalArrangement = Arrangement.SpaceAround,
            ) {
                for (tab in RootTab.entries) {
                    BottomTab(
                        tab = tab,
                        selected = currentTab == tab,
                        badge = if (tab == RootTab.SESSIONS) sessionCount else 0,
                        onTap = { onTabSelected(tab) },
                    )
                }
            }
        }
    }
}

@Composable
private fun BottomTab(
    tab: RootTab,
    selected: Boolean,
    badge: Int,
    onTap: () -> Unit,
) {
    val colors = appColors
    val color = if (selected) colors.accent else colors.textMuted

    Column(
        Modifier
            .clip(RoundedCornerShape(10.dp))
            .clickable(onClick = onTap)
            .padding(horizontal = 10.dp, vertical = 4.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Box(Modifier.width(42.dp).height(24.dp), contentAlignment = Alignment.Center) {
            Icon(
                if (selected) tab.selectedIcon else tab.icon,
                contentDescription = tab.label,
                tint = color,
                modifier = Modifier.size(23.dp),
            )
            if (badge > 0) {
                Box(
                    Modifier
                        .align(Alignment.TopEnd)
                        .padding(top = 0.dp)
                        .background(colors.accent, RoundedCornerShape(8.dp))
                        .padding(horizontal = 5.dp, vertical = 1.dp),
                ) {
                    Text(
                        "$badge",
                        style = TextStyle(
                            fontSize = 9.sp,
                            lineHeight = 11.sp,
                            fontWeight = FontWeight.W700,
                            color = colors.onAccent,
                        ),
                    )
                }
            }
        }
        Spacer(Modifier.height(4.dp))
        Text(
            tab.label,
            style = TextStyle(fontSize = 10.sp, fontWeight = FontWeight.W700, color = color),
        )
    }
}
